### DISCUSSION TITLE
How to Return a Binary File and JSON Metadata in the Same API Endpoint?

### DISCUSSION THREAD

**Discussion by kgtl12** [fastapi/fastapi#12983] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import FastAPI, UploadFile, File
from fastapi.responses import StreamingResponse
import json
import io

app = FastAPI()

@app.post("/upload/")
async def upload_and_return(file: UploadFile = File(...)):
    # Read file content
    file_content = await file.read()

    # Metadata as JSON
    metadata = {
        "file_name": file.filename,
        "file_type": file.content_type,
        "description": "This is a sample file"
    }

    # Return StreamingResponse with metadata in headers
    return StreamingResponse(
        io.BytesIO(file_content),
        media_type=file.content_type,
        headers={"metadata": json.dumps(metadata)}
    )

if __name__ == "__main__":
    import uvicorn
    uvicorn.run(app, host="0.0.0.0", port=8000)
```

### Description

I'm working on a FastAPI application where I need to return both a binary file (e.g., an image or PDF) and some associated metadata (in JSON format) in the same API response. 

What is the best practice to achieve this? Should I use a multipart response, include metadata in headers, or follow another approach? Here's a minimal script I've written that uses `StreamingResponse` for the file and JSON in the headers, but I’m open to suggestions for improvement or alternative solutions.

In this script, the file is streamed to the client, and the metadata is included in the headers as a JSON string. Is this a good approach, or are there better ways to handle such use cases? 

Looking forward to your feedback!

### Operating System

Linux

### FastAPI Version

0.109.2

### Pydantic Version

2.7.4

### Python Version

3.10.12

**Comment by luzzodev** [comment]
Hi @kgtl12,
here you have some considerations about your questions.

As always depends on what your use case is, there isn't a single magic implementation that fits all requirements.

What should be avoided for sure is to use headers to send part of the response content for a lot of reasons.

1. Is the file in response really required? if yes, can you provide a valid example?
2. There are two approaches that can solve your needs: one simpler and one more complex.
    -  The simpler one is by encoding the file and return it with metadata in a json response. This is a good solution if your files are not big, fits all in memory and the overhead on file size is acceptable.
    - Use a multipart-response (multipart/mixed). This could be a more solid approach but usually you should write the implementation because a lot of clients do not support it by default.

**Reply by kgtl12** [reply]
Hi @luzzodev ,

Yes the binary file is required. 

Example:
```python
class object_example(BaseModel):
    field1: str
    field2: int
   field3: float
   field4: dict
```
and a file content using `io` library or any other option to return a binary that might be bigger.
```python
    return StreamingResponse(
             io.BytesIO(file_content)
```

**Reply by luzzodev** [reply]
as far as your file is in memory the streaming response is quite useless

```
# Read file content
    file_content = await file.read()
```

Try first approach number 1 that is easier to implement then if memory, file size etc are a problem go with number2.

To encode the file you can use base64

```
# Encode binary data to base64
base64_encoded = base64.b64encode(binary_data).decode('utf-8')

# Create JSON object
json_data = {
    "filename": "example.bin",
    "file_data": base64_encoded
}

# Convert to JSON string
json_string = json.dumps(json_data)
```

and of course on client you have to the decode the file.

**Reply by kgtl12** [reply]
@luzzodev Thanks, but I am looking for a solution without converting to base64

**Reply by luzzodev** [reply, marked as answer]
then go with solution approach n2

**Reply by kgtl12** [reply]
@luzzodev great, thank you!
